compiler/src/iree/compiler/Dialect/Flow/Transforms/InterchangeTransposeGenericOps.cpp

```diff
@@ -0,0 +1,109 @@
+// Copyright 2022 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+//===--------------- InterchangeTransposeGenericOps.cpp -------------------===//
+//
+// Interchange loops in generic ops to make the transpose happen on the outputs
+// instead of inputs.
+//
+//===----------------------------------------------------------------------===//
+
+#include "iree/compiler/Dialect/Flow/Transforms/FusionUtils.h"
+#include "iree/compiler/Dialect/Flow/Transforms/PassDetail.h"
+#include "iree/compiler/Dialect/Flow/Transforms/Passes.h"
+#include "mlir/Dialect/Linalg/IR/Linalg.h"
+#include "mlir/Dialect/Linalg/Transforms/Transforms.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+namespace mlir {
+namespace iree_compiler {
+namespace IREE {
+namespace Flow {
+
+namespace {
+
+struct TransposeGenericOpPattern : public OpRewritePattern<linalg::GenericOp> {
+  using OpRewritePattern<linalg::GenericOp>::OpRewritePattern;
+  LogicalResult matchAndRewrite(linalg::GenericOp genericOp,
+                                PatternRewriter &rewriter) const override {
+    if (!genericOp.hasTensorSemantics()) return failure();
+
+    // indexing maps for linalg named ops
+    SmallVector<AffineMap> indexingMaps;
+
+    for (auto operand : genericOp.getInputOperands()) {
+      // check if the generic op has a non-identity map for the operand.
+      auto indexingMap = genericOp.getTiedIndexingMap(operand);
+      if (indexingMap.isIdentity()) {
+        // already in a good form.
+        return failure();
+      }
+
+      indexingMaps.push_back(indexingMap);
+
+      // check if the operation is fully parallel.
+      if ((indexingMap.getNumDims() != genericOp.getNumParallelLoops()) ||
+          !indexingMap.isPermutation())
+        return failure();
+
+      auto producer = operand->get().getDefiningOp<linalg::LinalgOp>();
+      if (!producer) continue;
+
+      if (isa<linalg::GenericOp>(producer)) return failure();
```

Comment:
I think we can drop this condition to. This is done post elementwise fusion. So any trivial fusion is already done. We can assume that now we are only looking for fusion using tile + fuse.